lib/internal/fs/watchers.js

```diff
@@ -270,10 +295,98 @@ ObjectDefineProperty(FSEvent.prototype, 'owner', {
   set(v) { return this[owner_symbol] = v; }
 });
 
+async function* watch(filename, options = {}) {
+  const path = toNamespacedPath(getValidatedPath(filename));
+  validateObject(options, 'options');
+
+  const {
+    persistent = true,
+    recursive = false,
+    encoding = 'utf8',
+    signal,
+  } = options;
+
+  validateBoolean(persistent, 'options.persistent');
+  validateBoolean(recursive, 'options.recursive');
+  validateAbortSignal(signal, 'options.signal');
+
+  if (encoding && !isEncoding(encoding)) {
+    const reason = 'is invalid encoding';
+    throw new ERR_INVALID_ARG_VALUE(encoding, 'encoding', reason);
+  }
+
+  if (signal?.aborted)
+    throw new AbortError();
+
+  const handle = new FSEvent();
+  let callback;
+  let rej;
+  const oncancel = () => {
+    handle.close();
+    rej(new AbortError());
+  };
+
+  try {
+    signal?.addEventListener('abort', oncancel, { once: true });
+
+    let promise = new Promise((resolve, reject) => {
+      callback = resolve;
+      rej = reject;
+    });
+
+    handle.onchange = (status, eventType, filename) => {
+      if (signal?.aborted) {
+        handle.close();
+        rej(new AbortError());
+        return;
+      }
+      if (status < 0) {
+        const error = uvException({
+          errno: status,
+          syscall: 'watch',
+          path: filename
+        });
+        error.filename = filename;
+        handle.close();
+        rej(error);
+        return;
+      }
+
+      callback({ eventType, filename });
+    };
+
+    const err = handle.start(path, persistent, recursive, encoding);
+    if (err) {
+      const error = uvException({
+        errno: err,
+        syscall: 'watch',
+        path: filename,
+        message: err === UV_ENOSPC ?
+          'System limit for number of file watchers reached' : ''
+      });
+      error.filename = filename;
+      handle.close();
+      throw error;
+    }
+
+    while (!signal?.aborted) {
+      yield await promise;
```

Comment:
I see a possible issue - if:

 - An event is emitted (but the user did not consume it yet).
 - Then an abort occurs (so the handle will close)
 - The onabort code will not actually reject anything (since the promise is already fulfilled)
 - The next time the user asks for events (keeps iterating the `for await` they get a forever pending promise)

Pretty easy to address but still - should probably be addressed :]

Reply:
A bit difficult to come up with a specific test for that but I do think it's covered actually. Did you have something more specific as a possible fix?

Reply:
I added a test but not sure if it exactly covers the potential edge case.

Reply:
Nevermind, I see there is a `throw new Error` below so the only difference here is that the user gets one "buffered" event because either the fs event already fired before they get their AbortError which sounds reasonable to me.